BacktraceScreen.c

```diff
@@ -0,0 +1,384 @@
+
+   char *completeFunctionName = NULL;
+   xAsprintf(&completeFunctionName, "%s+0x%zx", functionName, frame->offset);
+
+   char* objectDisplayed = frame->objectName;
+   size_t objectLength = printingHelper->maxObjNameLen;
+   if (!!(displayOptions & SHOW_FULL_PATH_OBJECT)) {
+      objectDisplayed = frame->objectPath;
+      objectLength = printingHelper->maxObjPathLen;
+   }
+
+   char* line = NULL;
+   int objectPathStart = -1;
+   int len = xAsprintf(&line, "%-*d 0x%0*zx %-*s %n%-*s",
+      (int)printingHelper->maxFrameNumLen, frame->index,
+      (int)printingHelper->maxAddrLen - (int)strlen("0x"), frame->address,
+      (int)maxFunctionNameLength, completeFunctionName,
+      &objectPathStart,
+      (int)objectLength, objectDisplayed ? objectDisplayed : "-"
+   );
+
+   int colors = CRT_colors[DEFAULT_COLOR];
+   if (!objectDisplayed) {
+      colors = CRT_colors[DYNAMIC_DARKGRAY];
+   }
+
+   RichString_appendnAscii(out, colors, line, len);
+
+   BacktraceFrame_highlightBasename(frame, out, line, objectPathStart);
+
+   free(completeFunctionName);
+   free(line);
+}
+
+static int BacktraceFrame_compare(const void* object1, const void* object2) {
+   const BacktraceFrame* frame1 = (const BacktraceFrame*)object1;
+   const BacktraceFrame* frame2 = (const BacktraceFrame*)object2;
+   return String_eq(frame1->functionName, frame2->functionName);
+}
+
+static void BacktraceFrame_delete(Object* object) {
+   BacktraceFrame* this = (BacktraceFrame*)object;
+   free(this->functionName);
+   free(this->demangleFunctionName);
+   free(this->objectPath);
+   free(this->objectName);
+   free(this);
+}
+
+static void BacktracePanel_setError(BacktracePanel* this, const char* error) {
+   assert(error);
+   assert(!this->error);
+
+   Panel* super = &this->super;
+   Panel_prune(super);
+
+   Vector* lines = this->super.items;
+   if (lines) {
+      Vector_delete(lines);
+   }
+
+   this->super.items = Vector_new(Class(ListItem), true, DEFAULT_SIZE);
+   Panel_set(super, 0, (Object*)ListItem_new(error, 0));
+}
+
+static void BacktracePanel_makePrintingHelper(const BacktracePanel* this, BacktracePanelPrintingHelper* printingHelper) {
+   Vector* lines = this->super.items;
+   size_t longestAddress = 0;
+
+   for (int i = 0; i < Vector_size(lines); i++) {
+      BacktraceFrame* frame = (BacktraceFrame*)Vector_get(lines, i);
+      frame->backtracePanel = this;
+
+      size_t digitOfOffsetFrame = strlen("+0x") + countDigit(frame->offset, 16);
+      if (frame->demangleFunctionName) {
+         size_t demangledFunctionNameLength = strlen(frame->demangleFunctionName) + digitOfOffsetFrame;
+         printingHelper->maxDemangledFuncNameLen = MAXIMUM(demangledFunctionNameLength, printingHelper->maxDemangledFuncNameLen);
+      }
+
+      if (frame->functionName) {
+         size_t functionNameLength = strlen(frame->functionName) + digitOfOffsetFrame;
+         printingHelper->maxFuncNameLen = MAXIMUM(functionNameLength, printingHelper->maxFuncNameLen);
```

Comment:
What should happen if both `frame->functionName` and `frame->demangleFunctionName` are NULL? It seems that `digitOfOffsetFrame` would be unused. Was that intentional?

Reply:
Yes, I think the code is more readable than checking if `frame->functionName` and `frame->demangleFunctionName` are not null, then checking if `frame->functionName` is not null, then `frame->demangleFunctionName` is not null.